File: frontend/src/services/logger.ts
import { Logger } from '../types';

class ConsoleLogger implements Logger {
    private formatMessage(level: string, message: string, data?: any): string {
        const timestamp = new Date().toISOString();
        const dataString = data ? JSON.stringify(data, null, 2) : '';
        return `[${timestamp}] [${level}] ${message} ${dataString}`;
    }

    info(message: string, data?: any): void {
        console.info(this.formatMessage('INFO', message, data));
    }

    error(message: string, error: any): void {
        console.error(this.formatMessage('ERROR', message, error));
    }

    warn(message: string, data?: any): void {
        console.warn(this.formatMessage('WARN', message, data));
    }

    debug(message: string, data?: any): void {
        console.debug(this.formatMessage('DEBUG', message, data));
    }
}

export const logger = new ConsoleLogger(); 

File: backend/app/core/logger.py
import logging
import sys
from logging.handlers import RotatingFileHandler
from pathlib import Path
import json
from datetime import datetime

class StructuredLogger:
    def __init__(self, name: str):
        self.logger = logging.getLogger(name)
        self.logger.setLevel(logging.DEBUG)
        
        # Create logs directory if it doesn't exist
        log_dir = Path("logs")
        log_dir.mkdir(exist_ok=True)
        
        # File handler with rotation
        file_handler = RotatingFileHandler(
            log_dir / "grc_chatbot.log",
            maxBytes=10*1024*1024,  # 10MB
            backupCount=5
        )
        file_handler.setLevel(logging.DEBUG)
        
        # Console handler
        console_handler = logging.StreamHandler(sys.stdout)
        console_handler.setLevel(logging.INFO)
        
        # Create formatters
        file_formatter = logging.Formatter(
            '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
        )
        console_formatter = logging.Formatter(
            '%(asctime)s - %(levelname)s - %(message)s'
        )
        
        file_handler.setFormatter(file_formatter)
        console_handler.setFormatter(console_formatter)
        
        self.logger.addHandler(file_handler)
        self.logger.addHandler(console_handler)
    
    def _format_message(self, message: str, **kwargs) -> str:
        """Format message with additional context as JSON"""
        log_data = {
            "message": message,
            "timestamp": datetime.utcnow().isoformat(),
            **kwargs
        }
        return json.dumps(log_data)
    
    def debug(self, message: str, **kwargs):
        self.logger.debug(self._format_message(message, **kwargs))
    
    def info(self, message: str, **kwargs):
        self.logger.info(self._format_message(message, **kwargs))
    
    def warning(self, message: str, **kwargs):
        self.logger.warning(self._format_message(message, **kwargs))
    
    def error(self, message: str, **kwargs):
        self.logger.error(self._format_message(message, **kwargs))
    
    def critical(self, message: str, **kwargs):
        self.logger.critical(self._format_message(message, **kwargs))

# Create logger instances for different components
api_logger = StructuredLogger("api")
service_logger = StructuredLogger("service")
config_logger = StructuredLogger("config") 